src/charm.py

```diff
@@ -58,11 +78,45 @@ def _on_peer_relation_changed(self, event):
             ] = event.relation.data[event.app]["MYSQL_ROOT_PASSWORD"]
             logger.info("Storing MYSQL_ROOT_PASSWORD in StoredState")
 
-    def _on_config_changed(self, _):
-        """This method handles the .on.config_changed() event"""
-        self._configure_pod()
+    def _on_config_changed(self, event):
+        """Set a new Juju pod specification"""
+        self.unit.status = MaintenanceStatus("Setting up containers.")
+        needs_restart = False
+        container = self.unit.get_container(PEER)
+        services = container.get_plan().to_dict().get("services", {})
+
+        try:
+            layer = self._mysql_layer()
+        except MySQLRootPasswordError as e:
+            logger.debug(e)
+            event.defer()
+            return
+
+        if (
+            not services
+            or services[PEER]["environment"]
+            != layer["services"][PEER]["environment"]
+        ):
+            container.add_layer(PEER, layer, combine=True)
+            needs_restart = True
+
+        if needs_restart:
+            try:
+                service = container.get_service("mysql")
+            except ConnectionError:
+                logger.info("Pebble API is not yet ready")
+                return
+            except ModelError:
+                logger.info("MySQL service is not yet ready")
+                return
+
+            if service.is_running():
+                container.stop("mysql")
+
+            container.start("mysql")
+            logger.info("Restarted MySQL service")
```

Comment:
I would break this `_on_config_changed` into two functions:
- `update_layer() -> bool`, which return True if changes we made, and
- `restart_service()`, which is called if changes were made.

Reply:
Yes! I was thinking the same!
done